docs/DONOTMERGE/pr-402.md

```diff
@@ -0,0 +1,108 @@
+      As of UTxO HD, the major portions of those two ledger states would be on disk instead of memory, but there'd still need to be two copies (later development could _maybe_ allow sharing the disk part).
+    - To handle rollbacks, the client would also need to maintain its own copy of Consensus's `LedgerDB` abstraction.
+    - (The [`foldBlocks`](https://input-output-hk.github.io/cardano-node/cardano-api/lib/Cardano-Api-LedgerState.html#v:foldBlocks) approach does this, but only for the immutable chain.)
+
+- Piggy-back on a `db-sync` instance.
+    - REQUIREMENT VIOLATION: This violates the desideratum regarding holding second copy of the ledger state in memory.
+    - @kderme remarked that he and @jasagredo charted a path towards using UTxO HD to ameliorate the memory cost, as sketched in the ChainSync option above.
+
+- Rely on PR #402's `rnHandleLedgerEvent` et al.
+    - REQUIREMENT VIOLATION: This violates the desideratum about becoming synced with a node.
+      The only workaround at the moment is to restart the node and force it to re-apply whichver historical blocks the client needs to process.
+      Moreover, currently, the only way to force the node to do so is to delete certains parts of its on-disk state!
+    - To handle rollbacks, the client would also need to maintain something akin to Consensus's `LedgerDB` abstraction, but more lightweight since its elements need not be entire Cardano ledger states.
+    - @kderme also remarked that `db-sync` only uses the ledger events for so much, and so still relies on the ledger state itself for the rest; this suggests that the blocks and their ledger events alone might not suffice for clients.
+
+The envisioned longer-term solutions are as follows.
+
+- Enrich the node to permanently store ledger events alongside blocks.
+    - Then the ChainSync approach above could simply serve ledger events alongside the blocks.
+    - This solution would very explicitly violate the design principles of the node, since this is non-trivial logic in the node that is absolutely not required for the core functionality of the blockchain itself.
+    - Also: what happens if the specification of ledger events for some block changes (eg a new ledger event is added)?
+
+- If a client is already maintaing the UTxO map (as does `db-sync`), then the release of UTxO HD might allow that app to avoid the UTxO-related heap costs of the ledger state---it could implement its own UTxO HD `BackingStore`.
+    - REQUIREMENT VIOLATION: This can only be implemented in Haskell.
+
+- Release a minimal proxy that uses the existing ChainSync approach above in order to maintain a database mapping blocks (ie slot and hash) to ledger events.
+    - This technically violates the second copy of ledger state in memory requirement.
+      However, the database should be reused among clients, which would amortize that RAM cost.
```

Comment:
There is an argument to be made about the second copy of ledger, which applies not only to this solution, but to all related. In the context of UTxO-HD and for a client that only reapplies blocks, its version of the UTxO doesn't need to be complete. In principle and if we're lucky in code already, calculating the ledger state and the events only ever needs the stake credentials of the UTxO.

I believe this observation needs further investigation with the ledger team.

Reply:
> calculating the ledger state and the events only ever needs the stake credentials of the UTxO

This seems incorrect me. Eventually, calculating a ledger state requires traversing the entire UTxO map in order to compute the new stake distribution, in order to compute the upcoming leader schedule. I don't see a way around that.

Maybe I'm overlooking something you implied with the "In the context of UTxO-HD" qualifier on your statement?

Reply:
The key observation here is that it needs to compute the **stake** distribution, which implies the need only of the staking part.